transformer.ts

```diff
@@ -0,0 +1,157 @@
+      let candidateName = `l10n\$${key.replace(/[^a-z0-9]+/g, '_')}`
+      for (
+        ;
+        l10nVariableNames[candidateName] &&
+        l10nVariableNames[candidateName] !== key;
+        candidateName += '_'
+      ) {
+        // This variable name is already used by another key.
+        // We need to find a new one.
+      }
+      l10nVariableNames[candidateName] = key
+      return candidateName
+    }
+
+    const visitor: ts.Visitor<ts.Node, ts.Node> = (node) => {
+      const match = ts.isCallExpression(node) && matchL10nNode(node)
+      if (match) {
+        return transformationContext.factory.updateCallExpression(
+          node,
+          node.expression,
+          undefined,
+          node.arguments.map((arg) => {
+            if (match.includes(arg)) {
+              if (!ts.isStringLiteral(arg)) {
+                return ts.visitEachChild(arg, visitor, transformationContext)
+              }
+              const key = arg.text
+              const variableName = getVariableName(key)
+              const variableReference =
+                transformationContext.factory.createIdentifier(variableName)
+
+              if (!l10nObjects[variableName]) {
+                const properties: ts.ObjectLiteralElementLike[] = []
+                for (const language of Object.keys(options.l10nData)) {
+                  const foundText = options.l10nData[language][arg.text]
+                  if (foundText && typeof foundText === 'string') {
+                    const stringLiteralNode =
+                      transformationContext.factory.createStringLiteral(
+                        foundText
+                      )
+                    ts.setEmitFlags(
+                      stringLiteralNode,
+                      ts.EmitFlags.NoAsciiEscaping
+                    )
+                    properties.push(
+                      transformationContext.factory.createPropertyAssignment(
+                        language,
+                        stringLiteralNode
+                      )
+                    )
+                  }
+                }
+                const object =
+                  transformationContext.factory.createObjectLiteralExpression(
+                    properties
+                  )
+                l10nObjects[variableName] = object
+              }
+              return transformationContext.factory.createCallExpression(
+                transformationContext.factory.createPropertyAccessExpression(
+                  transformationContext.factory.createIdentifier('__'),
+                  '$'
+                ),
+                [],
+                [arg, variableReference]
+              )
+            } else {
+              return ts.visitEachChild(arg, visitor, transformationContext)
+            }
+          })
+        )
+      }
+      return ts.visitEachChild(node, visitor, transformationContext)
+    }
+
+    const transformed = ts.visitNode(sourceFile, visitor) as ts.SourceFile
```

Comment:
Note to reviewers:
It is safe to use `as ts.SourceFile` here since `visitor` function returns the same kind (`ts.SourceFile`).

Without this type casting, `updateSourceFile` function will report a compilation error as it only accepts `ts.SourceFile` but not `ts.Node`.

Please note that `ts.Node` is like a super class of `ts.SourceFile`. So all `ts.SourceFile` is `ts.Node` but not the other way around.